distributed/protocol/numpy.py

```diff
@@ -23,7 +23,10 @@ def itemsize(dt):
 def serialize_numpy_ndarray(x):
     if x.dtype.hasobject:
         header = {"pickle": True}
-        frames = [pickle.dumps(x)]
+        frames = [None]
+        buffer_callback = lambda f: frames.append(memoryview(f))
```

Comment:
Why are we converting into a `memoryview` here?

Reply:
The `buffer_callback` is called on values of type `PickleBuffer`. Here's an example:

```python
In [1]: import pickle                                                           

In [2]: import numpy as np                                                      

In [3]: a = np.array([np.arange(3), np.arange(4, 6)], dtype=object)             

In [4]: l = []; d = pickle.dumps(a, protocol=5, buffer_callback=l.append)       

In [5]: l                                                                       
Out[5]: [<pickle.PickleBuffer at 0x115a64740>, <pickle.PickleBuffer at 0x115a647c0>]
```

Though `PickleBuffer` is a wrapper that expects to be passed some [`bytes`-like object]( https://docs.python.org/3.8/glossary.html#term-bytes-like-object ), which it wraps zero-copy. In the example above these are the nested NumPy arrays.

However as the rest of the serialization code expects frames to be `bytes` or `memoryview`s. We would either need to retool it to support `PickleBuffer` (with special casing when it is not available) or we can simply convert it to a `memoryview`, which is also zero-copy.

We've gone for the latter path of converting to a `memoryview` similar to what was done in PR ( https://github.com/dask/distributed/pull/3784 ). Note this is done without loss of generality.